### PR TITLE
fix(ivy): incorrectly generating shared pure function between null and object literal

### LINKED ISSUE #35298: Angular 9 confuses literal null with undefined when wrapped in object in input
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/.... I guess compiler

### Is this a regression?

Yes, the previous version in which this bug was not present was: 8.2.14

### Description

Angular 9 confuses literal `null` with `undefined` in inputs when the literal is wrapped in object and there are at least 2 instances of given component. In my project this breaks routing because there is a difference between

```html
<a [routerLink]="['', { outlets: { aux: null } }]"></a>
<a [routerLink]="['', { outlets: { aux: undefined } }]"></a>
```

## 🔬 Minimal Reproduction

1. execute following:

   ```sh
   git clone https://github.com/bgoscinski/angular-bug-repro1.git
   cd angular-bug-repro1
   npm install
   ng serve
   ```

1. observe the difference: 
image

1. see https://github.com/bgoscinski/angular-bug-repro1/blob/master/src/app/app.component.html#L17-L22

## 🌍 Your Environment

**Angular Version:**

<pre><code>
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.0.1
Node: 12.7.0
OS: win32 x64

Angular: 9.0.0
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.1
@angular-devkit/build-angular     0.900.1
@angular-devkit/build-optimizer   0.900.1
@angular-devkit/build-webpack     0.900.1
@angular-devkit/core              9.0.1
@angular-devkit/schematics        9.0.1
@angular/cli                      9.0.1
@ngtools/webpack                  9.0.1
@schematics/angular               9.0.1
@schematics/update                0.900.1
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
</code></pre>

**Anything else relevant?**

I guess the easiest workaround is to declare property on component instance and set its value to `null` and use it instead of literal `null` in templates

### PR DESCRIPTION
**Pull request by crisbeto**
In #33705 we made it so that we generate pure functions for object/array literals in order to avoid having them be shared across elements/views. The problem this introduced is that further down the line the `ContantPool` uses the generated literal in order to figure out whether to share an existing factory or to create a new one. `ConstantPool` determines whether to share a factory by creating a key from the AST node and using it to look it up in the factory cache, however the key generation function didn't handle function invocations and replaced them with `null`. This means that the key for `{foo: pureFunction0(...)}` and `{foo: null}` are the same.

These changes rework the logic so that instead of generating a `null` key for function invocations, we generate a variable called `<unknown>` which shouldn't be able to collide with anything.

Fixes #35298.

### CODE REVIEW

File: packages/core/test/acceptance/pure_function_spec.ts [hunk 1]
```diff
@@ -559,5 +559,65 @@ describe('components using pure function instructions internally', () => {
...
+      const values = fixture.componentInstance.directives.map(directive => directive.value);
+
+      expect(values).toEqual([{foo: null}, {foo: []}]);
+    });
+
+    it('should not confuse function calls and null inside of a literal', () => {
+      @Component({
+        template: `
+          <div [dir]="{foo: null}"></div>
+          <div [dir]="{foo: getFoo()}"></div>
```
Comment (crisbeto) [inline packages/core/test/acceptance/pure_function_spec.ts]: I think this one has been an issue even before #33705, but we just never hit it.

**Review by alxhub — APPROVED** [review]

File: packages/compiler/src/constant_pool.ts [hunk 2]
```diff
@@ -255,12 +256,25 @@ class KeyVisitor implements o.ExpressionVisitor {
     return `TYPEOF:${node.expr.visitExpression(this, context)}`;
   }
 
+  visitInvokeFunctionExpr(node: o.InvokeFunctionExpr, context: any): string {
+    return `FN_CALL:${node.fn.visitExpression(this, context)}:ARGS:${node.args.length}`;
+  }
+
+  visitInvokeMethodExpr(node: o.InvokeMethodExpr, context: any): string {
+    const receiver = node.receiver.visitExpression(this, context);
+    return `METHOD_CALL:${receiver}${node.name}:ARGS:${node.args.length}`;
```
Comment (alxhub) [inline packages/compiler/src/constant_pool.ts]: Should this not be
```suggestion
    return `METHOD_CALL:${receiver}:${node.name}:ARGS:${node.args.length}`;
```

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/compiler/src/constant_pool.ts (hunk 2 above)
Comment (crisbeto) [reply packages/compiler/src/constant_pool.ts]: I think I meant that one to be `${receiver}.${node.name}`.

File: packages/compiler/src/constant_pool.ts [hunk 3]
```diff
@@ -127,16 +127,17 @@ export class ConstantPool {
...
+      const argumentsForKey =
+          literal.entries.map(e => e.isConstant() ? e : o.variable('<unknown>'));
       const key = this.keyOf(o.literalArr(argumentsForKey));
       return this._getLiteralFactory(key, literal.entries, entries => o.literalArr(entries));
     } else {
       const expressionForKey = o.literalMap(
           literal.entries.map(e => ({
                                 key: e.key,
-                                value: e.value.isConstant() ? e.value : o.literal(null),
+                                value: e.value.isConstant() ? e.value : o.variable('<unknown>'),
```
Comment (alxhub) [inline packages/compiler/src/constant_pool.ts]: Maybe extract the `o.variable(...)` to a constant, and add a comment about what's going on there.

**Review by alxhub — APPROVED** [review]
Please update the commit message as well - it's now out of date since the changes we made yesterday :-)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:296245846:BASE:296245880:1582224164789:469b435e)

**Review by kara — APPROVED** [review]
LGTM. Thanks for the detailed commit message! :)

**Pull request closed without merging**